### PR TITLE
Add support for Iv2-prefixed Client IDs into Github App Credential

### PR DESCRIPTION
**Pull request by stevensonmichel**
As Github encourages more the use of Client IDs, one can create a Github App Installation Lookup App Credential using Github Client IDs starting with Iv1. However, using Iv2 will throw an error.

Fixes: as Github is providing Client IDs starting with Iv2, Credentials should be able to be created with it. The changes in this PR adds support for Iv2, enabling Credentials to be created using them.

### CODE REVIEW

File: src/awx_plugins/credentials/github_app.py [hunk 1]
```diff
@@ -139,8 +139,15 @@ def _is_intish(app_id_candidate: str | int) -> bool:
 
 
 def _is_client_id(client_id_candidate: str) -> bool:
-    client_id_prefix = 'Iv1.'
-    if not client_id_candidate.startswith(client_id_prefix):
+    # Support both Iv1 and Iv2 client ID prefixes
+    valid_prefixes = ('Iv1.', 'Iv2.')
```
Comment (arrestle) [inline src/awx_plugins/credentials/github_app.py]: I would recommend allowing "Iv#." Since you don't want to break again if they come up with Iv3. next year.

**Review by arrestle — DISMISSED** [review]
Great start! Python has https://docs.python.org/3/library/re.html or https://www.w3schools.com/python/python_regex.asp

File: src/awx_plugins/credentials/github_app.py (hunk 1 above)
Comment (stevensonmichel) [reply src/awx_plugins/credentials/github_app.py]: Sure

File: tests/github_app_test.py [hunk 2]
```diff
@@ -279,8 +297,22 @@ def token(self: '_FakeAppInstallationAuth') -> str:
...
+            'Iv23likIfIXeZTb5GBUU',
+            marks=github_app_jwt_client_id_unsupported,
+        ),
+    ),
+    ids=(
+        'app-id-int',
+        'app-id-str',
+        'client-id-iv1',
+        'client-id-iv2',
+        'client-id-iv23',
```
Comment (webknjaz) [inline tests/github_app_test.py]: These IDs don't explain what the respective tests are supposed to check. Could you make it clearer?

File: src/awx_plugins/credentials/github_app.py [hunk 3]
```diff
@@ -132,28 +132,23 @@ class EmptyKwargs(TypedDict):
...
 
 def _is_client_id(client_id_candidate: str) -> bool:
-    client_id_prefix = 'Iv1.'
-    if not client_id_candidate.startswith(client_id_prefix):
-        return False
-
-    client_id_trailer = client_id_candidate.removeprefix(client_id_prefix)
-
-    if len(client_id_trailer) != GH_CLIENT_ID_TRAILER_LENGTH:
-        return False
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py:149]: It might be a good idea to have two validation paths — for v1 and for v2. Then, v1 could remain stricter.

File: tests/github_app_test.py [hunk 4]
```diff
@@ -165,7 +173,17 @@ class AppInstallIds(TypedDict):
...
+        ),
+        pytest.param(
+            {
+                'app_or_client_id': 'Iv2.cccccccccccccccc',
+                'install_id': 'invalid',
+            },
+            '^Expected GitHub App Installation ID to be an integer '
+            "but got 'invalid'$",
+            id='gh-app-invalid-install-id-with-client-id-iv2',
             marks=github_app_jwt_client_id_unsupported,
```
Comment (webknjaz) [inline tests/github_app_test.py]: Is this even needed given that the test checks the install ID and doesn't care what's in the app/client ID? The additiona seems redundant.

File: tests/github_app_test.py [hunk 5]
```diff
@@ -165,7 +173,17 @@ class AppInstallIds(TypedDict):
             },
             '^Expected GitHub App Installation ID to be an integer '
             "but got 'invalid'$",
-            id='gh-app-invalid-install-id-with-client-id',
+            id='gh-app-invalid-install-id-with-client-id-iv1',
+            marks=github_app_jwt_client_id_unsupported,
```
Comment (webknjaz) [inline tests/github_app_test.py]: Following the linked issue chain, it seems that it might work now. It would be worth checking if anything changed before makring another test with xfail.

File: tests/github_app_test.py (hunk 2 above)
Comment (webknjaz) [reply tests/github_app_test.py]: Additionally, with the list of tests becoming longer, it's probably a good idea to move IDs into `pytest.param()` definitions and keep them coupled. It's okay for 2-3 tests but it'd difficult to maintain two coupled iterables when they are long.

**Comment by webknjaz** [comment]
@stevensonmichel FYI adding merge commits into PR branches creates foxtrots, which is why using rebase is preferred.

File: src/awx_plugins/credentials/github_app.py [hunk 6]
```diff
@@ -156,6 +158,31 @@ def _is_client_id(client_id_candidate: str) -> bool:
     return True
 
 
+def _is_iv1_prefix(client_id_candidate: str) -> bool:
+    """Check if it starts with Iv1 (not Iv10, Iv100, etc.)."""
+    if not client_id_candidate.startswith('Iv1'):
+        return False
+    if len(client_id_candidate) <= 3:
+        return True
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]: Why is this true?

File: src/awx_plugins/credentials/github_app.py [hunk 7]
```diff
@@ -132,18 +132,20 @@ class EmptyKwargs(TypedDict):
 
 GH_CLIENT_ID_TRAILER_LENGTH = 16
 HEXADECIMAL_BASE = 16
+GH_CLIENT_ID_MIN_LENGTH = 19
+GH_CLIENT_ID_MAX_LENGTH = 22
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]: Let's try keeping coupled values together:
```suggestion
GH_CLIENT_ID_LENGTH_BOUNDS = range(19, 23)  # min 19, max 22
```

File: src/awx_plugins/credentials/github_app.py [hunk 8]
```diff
@@ -156,6 +158,31 @@ def _is_client_id(client_id_candidate: str) -> bool:
...
+
+    if _is_iv1_prefix(client_id_candidate):
+        return _is_iv1_client_id(client_id_candidate)
+
+    # For Iv2+, just check length
+    return (
+        client_id_candidate.startswith('Iv')
+        and GH_CLIENT_ID_MIN_LENGTH
+        <= len(client_id_candidate)
+        <= GH_CLIENT_ID_MAX_LENGTH
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]:
```suggestion
        and len(client_id_candidate) in GH_CLIENT_ID_LENGTH_BOUNDS
```

**Review by webknjaz — COMMENTED** [review]
I feel like keeping a range in the same object is a bit more readable.

File: src/awx_plugins/credentials/github_app.py (hunk 3 above)
Comment (stevensonmichel) [reply src/awx_plugins/credentials/github_app.py:149]: Alright! Resolved

File: src/awx_plugins/credentials/github_app.py (hunk 6 above)
Comment (stevensonmichel) [reply src/awx_plugins/credentials/github_app.py]: This is just to prevent and indexError crash. A suitable example would be a client id with only 3 characters

File: src/awx_plugins/credentials/github_app.py (hunk 8 above)
Comment (stevensonmichel) [reply src/awx_plugins/credentials/github_app.py]: Fixed

File: src/awx_plugins/credentials/github_app.py (hunk 7 above)
Comment (stevensonmichel) [reply src/awx_plugins/credentials/github_app.py]: Fixed

File: tests/github_app_test.py (hunk 2 above)
Comment (stevensonmichel) [reply tests/github_app_test.py]: Fixed

File: src/awx_plugins/credentials/github_app.py (hunk 6 above)
Comment (webknjaz) [reply src/awx_plugins/credentials/github_app.py]: Turned out this is confusing to parse because past the first guard expression, all values of `client_id_candidate` would start with `Iv1` which is 3 exactly. And checking for less than 3 is counterintuitive.

File: src/awx_plugins/credentials/github_app.py [hunk 9]
```diff
@@ -156,6 +157,29 @@ def _is_client_id(client_id_candidate: str) -> bool:
     return True
 
 
+def _is_iv1_prefix(client_id_candidate: str) -> bool:
+    """Check if it starts with Iv1 (not Iv10, Iv100, etc.)."""
+    if not client_id_candidate.startswith('Iv1'):
+        return False
+    if len(client_id_candidate) <= 3:
+        return True
+    return not client_id_candidate[3].isdigit()
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]:
```suggestion
    known_v1_prefix = 'Iv1'

    if not client_id_candidate.startswith(known_v1_prefix):
        return False

    client_id_candidate_trailer = client_id_candidate.removeprefix(known_v1_prefix)

    if not client_id_candidate_trailer:  # empty trailer, no digits to check
        return True

    return not client_id_candidate_trailer[0].isdigit()
```

File: src/awx_plugins/credentials/github_app.py [hunk 10]
```diff
@@ -156,6 +157,29 @@ def _is_client_id(client_id_candidate: str) -> bool:
     return True
 
 
+def _is_iv1_prefix(client_id_candidate: str) -> bool:
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]: Looking closer, I don't understand why you don't check for `Iv1.`

File: src/awx_plugins/credentials/github_app.py [hunk 11]
```diff
@@ -156,6 +157,29 @@ def _is_client_id(client_id_candidate: str) -> bool:
...
+        return True
+    return not client_id_candidate[3].isdigit()
+
+
+def _is_client_id(client_id_candidate: str) -> bool:
+    # Iv1: strict format with Iv1. + 16 hex chars
+    # Iv2+: just check length (19-22 chars)
+
+    if _is_iv1_prefix(client_id_candidate):
+        return _is_iv1_client_id(client_id_candidate)
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]: Why can't this just be
```suggestion
    if _is_iv1_client_id(client_id_candidate):
        return True
```
?

File: src/awx_plugins/credentials/github_app.py [hunk 12]
```diff
@@ -132,18 +132,19 @@ class EmptyKwargs(TypedDict):
...
-def _is_client_id(client_id_candidate: str) -> bool:
-    client_id_prefix = 'Iv1.'
-    if not client_id_candidate.startswith(client_id_prefix):
+def _is_iv1_client_id(client_id_candidate: str) -> bool:
+    """Check if it's a valid Iv1 client ID (strict: Iv1. + 16 hex chars)."""
+    if not client_id_candidate.startswith('Iv1.'):
         return False
 
-    client_id_trailer = client_id_candidate.removeprefix(client_id_prefix)
+    client_id_trailer = client_id_candidate.removeprefix('Iv1.')
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]: No need to duplicate literals in multiple places. Revert this to the original forms.
```suggestion
    client_id_prefix = 'Iv1.'
    if not client_id_candidate.startswith(client_id_prefix):
        return False

    client_id_trailer = client_id_candidate.removeprefix(client_id_prefix)
```

File: src/awx_plugins/credentials/github_app.py [hunk 13]
```diff
@@ -156,6 +157,29 @@ def _is_client_id(client_id_candidate: str) -> bool:
...
 
+def _is_iv1_prefix(client_id_candidate: str) -> bool:
+    """Check if it starts with Iv1 (not Iv10, Iv100, etc.)."""
+    if not client_id_candidate.startswith('Iv1'):
+        return False
+    if len(client_id_candidate) <= 3:
+        return True
+    return not client_id_candidate[3].isdigit()
+
+
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]: With other observations, this becomes unnecessary.
```suggestion
```

File: src/awx_plugins/credentials/github_app.py [hunk 14]
```diff
@@ -132,18 +132,19 @@ class EmptyKwargs(TypedDict):
 
 GH_CLIENT_ID_TRAILER_LENGTH = 16
 HEXADECIMAL_BASE = 16
+GH_CLIENT_ID_LENGTH_BOUNDS = (19, 22)
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]: Let's unbreak this by applying the original suggestion I made in https://github.com/ansible/awx-plugins/pull/139#discussion_r2473386861
```suggestion
GH_CLIENT_ID_LENGTH_BOUNDS = range(19, 23)  # min 19, max 22
```

`thing in (1, 42)` will only be true if `thing == 1 or thing == 42`. Using the range allows anything in-between. Just take into account that the second argument is not included in the range. So `3 in range(1, 42) is True` and `41 in range(1, 42) is True`, but `42 in range(1, 42) is False`.

File: tests/github_app_test.py [hunk 15]
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
...
+        pytest.param(
+            {
+                'app_or_client_id': 'Iv23toolongtobevalidids',
+                'install_id': '666',
+            },
+            '^Expected GitHub App or Client ID to be an integer or a string '
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv23toolongtobevalidids'$",
+            id='gh-app-id-iv2-too-long',
```
Comment (webknjaz) [inline tests/github_app_test.py]: To catch bugs like the one I found in, https://github.com/ansible/awx-plugins/pull/139/files?new_files_changed=true#r2478577419, add tests for the lengths matching the boundaries — with 18/19/22/23 chars exactly.

File: tests/github_app_test.py (hunk 5 above)
Comment (webknjaz) [reply tests/github_app_test.py]: @stevensonmichel I'm still waiting for the results. Marking the thread as resolved does not mean that this is resolved. It's not.

File: tests/github_app_test.py [hunk 16]
```diff
@@ -273,19 +346,36 @@ def token(self: '_FakeAppInstallationAuth') -> str:
...
     ),
-    ids=('app-id-int', 'app-id-str', 'client-id'),
 )
 @pytest.mark.parametrize(
     'installation_id',
-    (456, '456'),
-    ids=('install-id-int', 'install-id-str'),
+    (
+        pytest.param(TEST_INSTALLATION_ID, id='install-id-int'),
+        pytest.param(str(TEST_INSTALLATION_ID), id='install-id-str-numeric'),
```
Comment (webknjaz) [inline tests/github_app_test.py]: I wouldn't touch unrelated places. This will end up extending the scope of the PR infinitely. It's a bad tone to smash together different types of changes (like refactoring/formatting/functional). Pick one and stick with it so that the PR is atomic and only contains changes that contribute to a single declared goal.

**Review by djyasin — APPROVED** [review]

File: src/awx_plugins/credentials/github_app.py (hunk 10 above)
Comment (arrestle) [reply src/awx_plugins/credentials/github_app.py]: because github switched to access tokens with a prefix of Iv2 with possible extensions to other prefixes... I'm not sure checking for ANY prefix makes sense any more. see https://docs.github.com/en/rest/apps/oauth-applications?apiVersion=2022-11-28

File: tests/github_app_test.py (hunk 15 above)
Comment (arrestle) [reply tests/github_app_test.py]: I would remove Iv1 and Iv2 - possibly just `^Iv[1-9]\d*\.[a-zA-Z0-9]+$` or use the endpoint 

`curl -L \
  -X POST \
  -H "Accept: application/vnd.github+json" \
  -u "<YOUR_CLIENT_ID>:<YOUR_CLIENT_SECRET>" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  "https://api.github.com/applications/<CLIENT_ID>/token" \
  -d '{"access_token":"<TOKEN_TO_VALIDATE>"}'`
  
See https://docs.github.com/en/rest/apps/oauth-applications?apiVersion=2022-11-28#check-a-token or https://docs.github.com/en/rest/apps/oauth-applications?apiVersion=2022-11-28

Comment (webknjaz) [reply tests/github_app_test.py]: Well, since there's already network I/O for retrieving the token, it will tell us if the creds are valid right in there. But it probably won't specify whether it's the client id or the token that's wrong.
If a regex was to be used, it'd probably need to differentiate between v1 and the rest. Though, I'm not a fan of regexes for such simple checks since they're more difficult to maintain. The coverage report won't tell you that you messed us the range boundaries in your regex, for example and it won't hint you how to cover unobvious cases.

File: src/awx_plugins/credentials/github_app.py (hunk 10 above)
Comment (webknjaz) [reply src/awx_plugins/credentials/github_app.py]: Not sure what that doc demonstrates. We already know that any validation is hunch-based. I'm actually starting to think that validation would be more useful in a form of a non-blocking hint rather than a thing that tries to predict whether input is valid. At least, for the inputs where GH doesn't document their exact shape.

File: src/awx_plugins/credentials/github_app.py [hunk 17]
```diff
@@ -156,6 +157,28 @@ def _is_client_id(client_id_candidate: str) -> bool:
...
 
 
+def _is_iv1_prefix(client_id_candidate: str) -> bool:
+    """Check if it starts with Iv1 (not Iv10, Iv100, etc.)."""
+    known_v1_prefix = 'Iv1'
+
+    if not client_id_candidate.startswith(known_v1_prefix):
+        return False
+
+    client_id_candidate_trailer = client_id_candidate.removeprefix(
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]: This should unbreak the CI
```suggestion
```

(*I think* it's GitHub's new experimental review interface that might've corrupted the patch)

File: src/awx_plugins/credentials/github_app.py [hunk 18]
```diff
@@ -132,18 +132,19 @@ class EmptyKwargs(TypedDict):
...
-def _is_client_id(client_id_candidate: str) -> bool:
-    client_id_prefix = 'Iv1.'
-    if not client_id_candidate.startswith(client_id_prefix):
+def _is_iv1_client_id(client_id_candidate: str) -> bool:
+    """Check if it's a valid Iv1 client ID (strict: Iv1. + 16 hex chars)."""
+    if not client_id_candidate.startswith('Iv1.'):
         return False
 
-    client_id_trailer = client_id_candidate.removeprefix(client_id_prefix)
+    client_id_trailer = client_id_candidate.removeprefix('Iv1.')
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]:
```suggestion
    client_id_prefix = 'Iv1.'
    if not client_id_candidate.startswith(client_id_prefix):
        return False

    client_id_trailer = client_id_candidate.removeprefix(client_id_prefix)
```

File: tests/github_app_test.py [hunk 19]
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
...
+                'app_or_client_id': 'Iv1',
                 'install_id': '666',
             },
             '^Expected GitHub App or Client ID to be an integer or a string '
-            r'starting with `Iv1\.` followed by 16 hexadecimal digits, but got'
-            " 'Iv1.bbbbbbbbbbbbbbb'$",
-            id='gh-app-id-client-id-not-enough-chars',
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv1'$",
```
Comment (webknjaz) [inline tests/github_app_test.py]:
```suggestion
            "but got 'Iv1'$",
```

File: tests/github_app_test.py [hunk 20]
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
...
+                'app_or_client_id': 'Iv1.bbb',
                 'install_id': '666',
             },
             '^Expected GitHub App or Client ID to be an integer or a string '
-            r'starting with `Iv1\.` followed by 16 hexadecimal digits, but got'
-            " 'Iv1.bbbbbbbbbbbbbbbx'$",
-            id='gh-app-id-client-id-broken-hex',
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv1.bbb'$",
```
Comment (webknjaz) [inline tests/github_app_test.py]:
```suggestion
            "but got 'Iv1.bbb'$",
```

File: tests/github_app_test.py [hunk 21]
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
...
+                'app_or_client_id': 'Iv1aaaaaaaaaaaaaaaa',
                 'install_id': '666',
             },
             '^Expected GitHub App or Client ID to be an integer or a string '
-            r'starting with `Iv1\.` followed by 16 hexadecimal digits, but got'
-            " 'Iv1.bbbbbbbbbbbbbbbbb'$",
-            id='gh-app-id-client-id-too-many-chars',
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv1aaaaaaaaaaaaaaaa'$",
```
Comment (webknjaz) [inline tests/github_app_test.py]:
```suggestion
            "but got 'Iv1aaaaaaaaaaaaaaaa'$",
```

File: tests/github_app_test.py [hunk 22]
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
...
+        ),
+        pytest.param(
+            {
+                'app_or_client_id': 'Iv1.gggggggggggggggg',
+                'install_id': '666',
+            },
+            '^Expected GitHub App or Client ID to be an integer or a string '
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv1.gggggggggggggggg'$",
```
Comment (webknjaz) [inline tests/github_app_test.py]:
```suggestion
            "but got 'Iv1.gggggggggggggggg'$",
```

File: tests/github_app_test.py [hunk 23]
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
...
+        ),
+        pytest.param(
+            {
+                'app_or_client_id': 'Iv1.aaaaaaaaaaaaaaaaa',
+                'install_id': '666',
+            },
+            '^Expected GitHub App or Client ID to be an integer or a string '
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv1.aaaaaaaaaaaaaaaaa'$",
```
Comment (webknjaz) [inline tests/github_app_test.py]:
```suggestion
            "but got 'Iv1.aaaaaaaaaaaaaaaaa'$",
```

File: tests/github_app_test.py [hunk 24]
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
...
+        # Iv2+ validation tests (relaxed: just length)
+        pytest.param(
+            {
+                'app_or_client_id': 'Iv',
+                'install_id': '666',
+            },
+            '^Expected GitHub App or Client ID to be an integer or a string '
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv'$",
```
Comment (webknjaz) [inline tests/github_app_test.py]:
```suggestion
            "but got 'Iv'$",
```

File: tests/github_app_test.py [hunk 25]
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
...
+        ),
+        pytest.param(
+            {
+                'app_or_client_id': 'Iv2.bbb',
+                'install_id': '666',
+            },
+            '^Expected GitHub App or Client ID to be an integer or a string '
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv2.bbb'$",
```
Comment (webknjaz) [inline tests/github_app_test.py]:
```suggestion
            "but got 'Iv2.bbb'$",
```

File: tests/github_app_test.py [hunk 26]
```diff
@@ -106,39 +108,100 @@ class AppInstallIds(TypedDict):
...
+        ),
+        pytest.param(
+            {
+                'app_or_client_id': 'Iv23toolongtobevalidids',
+                'install_id': '666',
+            },
+            '^Expected GitHub App or Client ID to be an integer or a string '
+            r'starting with `Iv1\.` \(16 hex chars\) or `Iv2\+` \(19-22 chars\), '
+            'but got'
+            " 'Iv23toolongtobevalidids'$",
```
Comment (webknjaz) [inline tests/github_app_test.py]:
```suggestion
            "but got 'Iv23toolongtobevalidids'$",
```

**Review by arrestle — DISMISSED** [review]
I would avoid relying on string comparison and use [this github endpoint](https://docs.github.com/en/rest/authentication/authenticating-to-the-rest-api?apiVersion=2022-11-28) to validate a token, if indeed this is even a requirement. 

The github token code may be in flux, and while currently seems to be Iv[1|2].[n+] future versions may not even follow that pattern.

File: src/awx_plugins/credentials/github_app.py [hunk 27]
```diff
@@ -132,28 +133,48 @@ class EmptyKwargs(TypedDict):
 
 GH_CLIENT_ID_TRAILER_LENGTH = 16
 HEXADECIMAL_BASE = 16
+GH_CLIENT_ID_MIN_LENGTH = 19
+GH_CLIENT_ID_MAX_LENGTH = 22
+GH_CLIENT_ID_LENGTH_BOUNDS = range(
+    GH_CLIENT_ID_MIN_LENGTH,
+    GH_CLIENT_ID_MAX_LENGTH + 1,
+)  # min 19, max 22
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]: Since you've reintroduced explicit constants, this comment is no longer necessary. In fact, there is harm in it potentially going out of sync and becoming confusing.
```suggestion
)
```

File: src/awx_plugins/credentials/github_app.py [hunk 28]
```diff
@@ -132,28 +133,48 @@ class EmptyKwargs(TypedDict):
...
 
 
 def _is_intish(app_id_candidate: str | int) -> bool:
     return isinstance(app_id_candidate, int) or app_id_candidate.isdigit()
 
 
 def _is_client_id(client_id_candidate: str) -> bool:
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]: Plz use a helper function instead of increasing indentation.

File: src/awx_plugins/credentials/github_app.py [hunk 29]
```diff
@@ -156,6 +163,28 @@ def _is_client_id(client_id_candidate: str) -> bool:
...
+    # Iv1 + non-digit (Iv1aaa): INVALID
+    # Iv2+: just check length (19-22 chars)
+
+    # Check for Iv1. with dot - strict validation
+    if _is_iv1_client_id(client_id_candidate):
+        return True
+
+    # Check for invalid Iv1 format: Iv1 followed by non-digit
+    # (not dot, not digit). Examples: Iv1aaa, Iv1xxx → INVALID
+    if re.match(r'^Iv1[^\d]', client_id_candidate):
```
Comment (webknjaz) [inline src/awx_plugins/credentials/github_app.py]: At this point, it should never be a v1 since `_is_iv1_client_id()` would be `False` here. But if we want this additional check, it belongs in that helper and shouldn't be in disconnected context.

**Comment by webknjaz** [comment]
In general, I agree with @arrestle in that we don't know anything about these tokens and might as well just drop any shape/length validation.

P.S. I'm still interested in learning if the underlying libraries are finally able to accept client IDs, though, and dropping `xfails` so it's actually tested somehow.

File: src/awx_plugins/credentials/github_app.py (hunk 29 above)
Comment (stevensonmichel) [reply src/awx_plugins/credentials/github_app.py]: Now, I see. As it is to validate cases like Iv10.... , I think my comment "Check for Iv1 format" is misleading. I will change that.

Comment (stevensonmichel) [reply src/awx_plugins/credentials/github_app.py]: Also, I tested PyGithub with client Ids, and it validated it return the token ghs_...

File: tests/github_app_test.py [hunk 30]
```diff
@@ -97,90 +93,9 @@ class AppInstallIds(TypedDict):
...
-    """Test that invalid arguments make token extractor bail early."""
-    with pytest.raises(ValueError, match=expected_error_msg):
-        gh_app_plugin_mod.extract_github_app_install_token(
-            github_api_url='https://github.com',
-            private_rsa_key='key',
-            **extract_github_app_install_token_args,
-        )
+# NOTE: Format validation tests removed - we no longer validate
+# Client ID formats upfront. GitHub's API is the ultimate validator.
+# Invalid credentials will be rejected by the API with appropriate errors.
```
Comment (webknjaz) [inline tests/github_app_test.py]: Make sure not to keep comments that aren't useful in files.

File: tests/github_app_test.py [hunk 31]
```diff
@@ -29,12 +29,8 @@
...
 
 
-github_app_jwt_client_id_unsupported = pytest.mark.xfail(
-    raises=(AssertionError, ValueError),
-    reason='Client ID in JWT is not currently supported by '
-    'PyGitHub and is disabled.\n\n'
-    'Ref: https://github.com/PyGithub/PyGithub/issues/3213',
-)
+# Client IDs are now supported - no longer using xfail marker
+# All Client ID tests should pass with the no-validation approach
```
Comment (webknjaz) [inline tests/github_app_test.py]: This detached statement is not needed. It doesn't add clarity, just confusion.

File: tests/github_app_test.py
Comment (webknjaz) [inline tests/github_app_test.py]: rename this to
```suggestion
    'application_or_client_id',
```

**Comment by webknjaz** [comment]
@stevensonmichel macos will stop failing once you rebase.

File: tests/github_app_test.py [hunk 32]
```diff
@@ -271,16 +177,16 @@ def token(self: '_FakeAppInstallationAuth') -> str:
...
         '123',
-        pytest.param(
-            'Iv1.aaaaaaaaaaaaaaaa',
-            marks=github_app_jwt_client_id_unsupported,
-        ),
+        'Iv1.aaaaaaaaaaaaaaaa',
+        'Iv2aaaaaaaaaaaaaaaa',
+        'Iv10aaaaaaaaaaaaaaaa',
+        'Iv100.aaaaaaaaaaaaaaaa',
+        'Iv23likIfIXeZTb5GCAA',
```
Comment (webknjaz) [inline tests/github_app_test.py]: These params become a part of the test ID. They show up in test reports and console output. It's not necessarily clear what they are for.
This is why, please wrap each entry with `pytest.param(..., id='human-oriented-test-case-identifier')` which would make it clear why these are being tested in the first place.
Don't use the separate `ids=` argument to `pytest.mark.parametrize()` since the list of things is too long for that style to be maintainable in this specific case.

File: src/awx_plugins/credentials/github_app.py (hunk 29 above)
Comment (webknjaz) [reply src/awx_plugins/credentials/github_app.py]: > I tested PyGithub with client Ids, and it validated it return the token ghs_...

@stevensonmichel good to know. I see that you've already taken care of the `xfail` markers so I'm marking this thread as resolved.

File: tests/github_app_test.py (hunk 31 above)
Comment (webknjaz) [reply tests/github_app_test.py]: Marking this thread as resolved.

File: tests/github_app_test.py
Comment (webknjaz) [reply tests/github_app_test.py]: Marking this thread as resolved.

File: tests/github_app_test.py (hunk 30 above)
Comment (webknjaz) [reply tests/github_app_test.py]: Marking this thread as resolved.

File: src/awx_plugins/credentials/github_app.py (hunk 18 above)
Comment (webknjaz) [reply src/awx_plugins/credentials/github_app.py]: Marking this thread as resolved.

File: tests/github_app_test.py (hunk 15 above)
Comment (webknjaz) [reply tests/github_app_test.py]: Marking this thread as resolved.

File: src/awx_plugins/credentials/github_app.py (hunk 13 above)
Comment (webknjaz) [reply src/awx_plugins/credentials/github_app.py]: Marking this thread as resolved.

File: src/awx_plugins/credentials/github_app.py (hunk 12 above)
Comment (webknjaz) [reply src/awx_plugins/credentials/github_app.py]: Marking this thread as resolved.

File: src/awx_plugins/credentials/github_app.py (hunk 11 above)
Comment (webknjaz) [reply src/awx_plugins/credentials/github_app.py]: Marking this thread as resolved.

File: src/awx_plugins/credentials/github_app.py (hunk 10 above)
Comment (webknjaz) [reply src/awx_plugins/credentials/github_app.py]: Marking this thread as resolved.

File: src/awx_plugins/credentials/github_app.py (hunk 9 above)
Comment (webknjaz) [reply src/awx_plugins/credentials/github_app.py]: Marking this thread as resolved.

File: src/awx_plugins/credentials/github_app.py (hunk 6 above)
Comment (webknjaz) [reply src/awx_plugins/credentials/github_app.py]: Marking this thread as resolved.

**Review by arrestle — APPROVED** [review]
Hi Stevenson! Good job getting this reviewed it's a tough one. 

I approve of the approach but honestly don't completely understand the implications of changing application_id to  application_or_client_id. However, I trust your team has enough integration testing to make sure this all works correctly! 

Let me know if you need any further assistance.

**Comment by stevensonmichel** [comment]
Thank you so much for the review and approval! I really appreciate your feedback. Regarding the application_id to application_or_client_id change, the name was updated to better reflect that it now accepts both GitHub App IDs and Client IDs. The key change in this PR is letting PyGitHub's API be the ultimate validator. This makes the plugin future-proof when GitHub introduces new ID formats.
